lib/AST/ASTMangler.cpp

```diff
@@ -2573,6 +2571,148 @@ void ASTMangler::appendSymbolicReference(SymbolicReferent referent) {
+//
+// - Sorts the inverse requirements in a canonical way.
+static void reconcileInverses(
+           SmallVector<InverseRequirement, 2> &inverses,
+           const DeclContext *baseContext,
+           std::optional<unsigned> inversesAlreadyMangledDepth) {
+  CanGenericSignature childSig;
+  if (baseContext)
+    if (auto sig = baseContext->getGenericSignatureOfContext())
+      childSig = sig.getCanonicalSignature();
+
+  if (childSig || inversesAlreadyMangledDepth)
+    llvm::erase_if(inverses, [&](InverseRequirement const& inv) -> bool {
+      // Drop inverses that aren't applicable in the nested / child signature,
+      // because of an added requirement.
+      if (childSig && childSig->requiresProtocol(inv.subject, inv.protocol))
+        return true;
+
+      auto gp = inv.subject->castTo<GenericTypeParamType>();
+
+      // Remove inverses that were either already mangled for this entity,
+      // or chosen not to be included in the output.
+      if (auto limit = inversesAlreadyMangledDepth)
+        if (gp->getDepth() <= limit)
+          return true;
+
+      return false;
+    });
+
+  // Sort inverse requirements for stability.
+  llvm::array_pod_sort(
+      inverses.begin(), inverses.end(),
+    [](const InverseRequirement *lhs, const InverseRequirement *rhs) -> int {
+      return lhs->compare(*rhs);
+    });
+}
+
+// If we're mangling an entity in a generic type context and it has inverse
+// requirements, then append the generic type and follow it with the mangling
+// of a constrained extension. We do this to preserve the invariant that
+// anything defined in an extension with the same signature as the nominal
+// will have the same mangling. In order to achieve this with inverse req's
+// we must mangle entities within type's who have inverse requirements to be
+// different than those in type's without them.
+//
+// For example:
+//
+//     struct X<Y: ~Copyable> {
+//       func foo() {} // In order for this to have the same mangling as the
+//                        one in the extension, we need to treat it as if it
+//                        were in the extension.
+//     }
+//
+//     extension X where Y: ~Copyable {
+//       func foo() {}
+//     }
+//
+// Thus, this function appends the remaining piece to an existing context to
+// make it look as if the context is an extension:
+//
+//     context ::= entity module generic-signature? 'E'
+//                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+//                        this piece will be appended, using the given args.
+//
+// Only the inverse requirements from the given signature will be included!
+void ASTMangler::appendContextualInverses(GenericSignature sig,
+                                          const ModuleDecl *module,
+                                          StringRef alternateModuleName) {
+  GenericSignatureParts parts;
+  gatherGenericSignatureParts(sig, /*contextSig=*/nullptr, parts);
+
+  if (parts.inverses.empty())
+    return;
```

Comment:
This early exit seems odd to me, because we're not mangling the module when we go this route. So either we don't need the early exit, or something else needs to ensure that the module gets mangled.

Reply:
Yeah this probably needs more explanation.

The trick here is that we're conditionally appending the extension, and want to avoid it entirely in the case when there's no inverse to mangle. By the point we call `appendContextualInverses`, we know that we've appended a context where `entity` is `nominal-type`:
```
context ::= entity
entity ::= nominal-type                    // named type declaration
```
Since extension mangling looks like this:
```
context ::= entity module generic-signature? 'E'
```
We want to avoid even appending the module.